### PR TITLE
Document new `compilerOptions` DSL for 2.1.0

### PR DESCRIPTION
**Pull request by sarahhaggarty**
This PR adds documentation for the newly stabilized `compilerOptions` DSL as part of Kotlin 2.1.0.

### CODE REVIEW

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 1]
```diff
@@ -828,12 +829,104 @@ kotlin {
...
+    }
+}
+```
+
+</tab>
+</tabs>
+
+### Compilation unit level
 
+To configure compiler options for a specific compilation, use the `compilerOptions {}` block inside the `compilations` object:
```
[the remark this answers is not in the document]
Comment (Tapchicoma) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:869]: Maybe better to mention a way to configure via task and then mention how to access task via `KotlinCompilation`.

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 2]
```diff
@@ -36,19 +36,13 @@ plugins {
...
-> The support for `compilerOptions {}` as a top-level block is [Experimental](components-stability.md#stability-levels-explained)
-> and requires opt-in. It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate
-> your feedback on it in [YouTrack](https://kotl.in/issue).
->
-{style="warning"}
+| **Block**            | **Description**                                                                                                                |
+|----------------------|--------------------------------------------------------------------------------------------------------------------------------|
+| _&lt;targetName&gt;_ | Declares a particular target of a project. The names of available targets are listed in the [Targets](#targets) section.       |
+| `targets`            | All targets of the project.                                                                                                    |
+| `presets`            | All predefined targets. Use this for [configuring multiple predefined targets](#targets) at once.                              |
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:43]: I removed documentation relating to deprecated `presets`

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:869]: I updated the code block. Can you check that it and the groovy version are correct now?
Can you also share kotlin/groovy samples for "a way to configure via task and then mention how to access task via KotlinCompilation"?

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 3]
```diff
@@ -783,29 +765,47 @@ kotlin {
...
 </tab>
 </tabs>
 
-Alternatively, to configure compiler options that are common for all targets, you can use the `compilerOptions {}` [top-level block](multiplatform-dsl-reference.md#top-level-blocks):
+## Compiler options
+
+You can configure compiler options in your projects at three different levels:
+
+* **Extension level**, by using it as a top-level block.
+* **Target level**.
```
Comment (Tapchicoma) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:782]: Maybe mention that it is available in the specific target DSL block?

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 4]
```diff
@@ -828,12 +827,106 @@ kotlin {
...
+}
+```
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+kotlin {
+    jvm {
+        compilations.main.compileTaskProvider.configure {
```
Comment (Tapchicoma) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:893]: This snippet works both for KTS and Groovy:
```
jvm {
        compilations.named(KotlinCompilation.MAIN_COMPILATION_NAME) {
            compileTaskProvider.configure {
                compilerOptions {
                    
                }
            }
        }
    }
```

Though your examples are working too, so up to you to change them.

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 5]
```diff
@@ -828,12 +827,106 @@ kotlin {
...
+</tab>
+</tabs>
+
+To configure compiler options for a specific task, use the `compilerOptions {}` block inside the task:
+
+<tabs group="build-script">
+<tab title="Kotlin" group-key="kotlin">
+
+```kotlin
+task.named<KotlinJvmCompile>("compileKotlinJvm"){
```
Comment (Tapchicoma) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:910]:
```suggestion
task.named<KotlinJvmCompile>("compileKotlinJvm") {
```

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 6]
```diff
@@ -828,12 +827,106 @@ kotlin {
...
+        allWarningsAsErrors.set(true)
+    }
+}
+```
+
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+task.named<KotlinJvmCompile>("compileKotlinJvm"){
```
Comment (Tapchicoma) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:921]:
```suggestion
task.named<KotlinJvmCompile>("compileKotlinJvm") {
```

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tapchicoma) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:867]: > Can you also share kotlin/groovy samples for "a way to configure via task and then mention how to access task via KotlinCompilation"?

Basically you `tasks.named` examples and then your examples with `compilations.named("main") { compileTaskProvider... }`

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 3 above)
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:782]: Would the below work?
> **Target level**, by using it in the target block.

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md]: Ok, I swapped the order!

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 3 above)
Comment (Tapchicoma) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:782]: yes

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 7]
```diff
@@ -828,12 +827,110 @@ kotlin {
...
+    compilerOptions {
+        allWarningsAsErrors = true
+    }
+}
+```
+
+</tab>
+</tabs>
+
+To configure compiler options for a specific compilation, use the `compilerOptions {}` block inside the `compilations` object:
```
Comment (Tapchicoma) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:894]: I would rephrase it to something like this:

You could also access compilation task provider from a specific compilation and configure this task compiler options via `compilerOptions {}` DSL

**Review by Tapchicoma — APPROVED** [review]

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 8]
```diff
@@ -36,24 +36,17 @@ plugins {
...
-> and requires opt-in. It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate
-> your feedback on it in [YouTrack](https://kotl.in/issue).
->
-{style="warning"}
+| **Block**            | **Description**                                                                                                                |
+|----------------------|--------------------------------------------------------------------------------------------------------------------------------|
+| _&lt;targetName&gt;_ | Declares a particular target of a project. The names of available targets are listed in the [Targets](#targets) section.       |
+| `targets`            | All targets of the project.                                                                                                    |
+| `sourceSets`         | Configures predefined and declares custom [source sets](#source-sets) of the project.                                          |
+| `compilerOptions`    | Extension-level common [compiler options](#compiler-options) that are used as defaults for all targets and shared source sets. |
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:44]:
```suggestion
| `compilerOptions`    | Specifies common extension-level [compiler options](#compiler-options) that are used as defaults for all targets and shared source sets. |
```

The word sequence "common extension-level options" sounds better to me than "extension-level common options" for some reason, but feel free to ignore that :)

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 9]
```diff
@@ -36,24 +36,17 @@ plugins {
...
 
 ## Targets
 
-_Target_ is a part of the build responsible for compiling, testing, and packaging a piece of software aimed at
-one of the supported platforms. Kotlin provides target presets for each platform. See how to [use a target preset](multiplatform-set-up-targets.md).
+A _target_ is a part of the build responsible for compiling, testing, and packaging a piece of software aimed at
+one of the supported platforms. Kotlin provides targets for each platform. See [Set up targets for Kotlin Multiplatform](multiplatform-set-up-targets.md).
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:49]: Is it just me, or is there a little bit of a non sequitur? "Kotlin provides targets for each platform. See how to set up targets." But I thought Kotlin provides them?..

Do you think we can include a transition here that would explain the use case, a reason to follow this link?

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 10]
```diff
@@ -134,20 +127,13 @@ Each target can have one or more [compilations](#compilations).
...
-
-> The support for `compilerOptions {}` as a common target configuration is [Experimental](components-stability.md#stability-levels-explained)
-> and requires opt-in. It may be dropped or changed at any time. Use it only for evaluation purposes. We would appreciate
-> your feedback on it in [YouTrack](https://kotl.in/issue).
->
-{style="warning"}
+| **Name**            | **Description**                                                                                                                                                                                | 
+|---------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `attributes`        | Attributes used for [disambiguating targets](multiplatform-set-up-targets.md#distinguish-several-targets-for-one-platform) for a single platform.                                              |
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:132]:
```suggestion
| `attributes`        | Attributes used to [disambiguate targets](multiplatform-set-up-targets.md#distinguish-several-targets-for-one-platform) for a single platform.                                              |
```

Also, the link text says "disambiguate" and the title of the section says "distinguish". Should they be aligned?

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 11]
```diff
@@ -134,20 +127,13 @@ Each target can have one or more [compilations](#compilations).
...
+| `platformType`      | Designates the Kotlin platform of this target. Available values: `jvm`, `androidJvm`, `js`, `wasm`, `native`, `common`.                                                                        |
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:133]:
```suggestion
| `platformType`      | The Kotlin platform for this target. Available values: `jvm`, `androidJvm`, `js`, `wasm`, `native`, `common`.                                                                        |
```

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 12]
```diff
@@ -134,20 +127,13 @@ Each target can have one or more [compilations](#compilations).
...
+| `artifactsTaskName` | The name of the task that builds the resulting artifacts of this target.                                                                                                                       |
+| `components`        | The components used to setup Gradle publications.                                                                                                                                              |
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:135]:
```suggestion
| `components`        | Components used to set up Gradle publications.                                                                                                                                              |
```

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 13]
```diff
@@ -134,20 +127,13 @@ Each target can have one or more [compilations](#compilations).
...
+| `compilerOptions`   | The [compiler options](#compiler-options) used for the target. This declaration overrides any `compilerOptions {}` configured at [top level](multiplatform-dsl-reference.md#top-level-blocks). |
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:136]:
```suggestion
| `compilerOptions`   | [Compiler options](#compiler-options) used for the target. This declaration overrides any `compilerOptions {}` configured at the [top level](multiplatform-dsl-reference.md#top-level-blocks). |
```

File: docs/topics/multiplatform/multiplatform-dsl-reference.md [hunk 14]
```diff
@@ -783,29 +765,47 @@ kotlin {
...
 
 </tab>
 </tabs>
 
-Alternatively, to configure compiler options that are common for all targets, you can use the `compilerOptions {}` [top-level block](multiplatform-dsl-reference.md#top-level-blocks):
+## Compiler options
+
+You can configure compiler options in your projects at three different levels:
+
+* **Extension level**, by using it as a top-level block.
```
Comment (zamulla) [inline docs/topics/multiplatform/multiplatform-dsl-reference.md:781]: I was a bit lost trying to understand what "it" refers to here. We could mention the `compilerOptions {}` block at least in the first list item, to smooth the experience.

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 13 above)
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:136]: I accepted half of this one. I didn't include "_the_ top level". It sounds more natural without "the" as there is only one top level.

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 14 above)
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:781]: What about this instead:
* **Extension level**, as a top-level block.
* **Target level**, in the target block.
* **Compilation unit level**, usually as a specific compilation task.
Then we can keep the structure lean and explain exactly _how_ (with the `compilerOptions {}` block) later in the section. WDYT?

File: docs/topics/multiplatform/multiplatform-dsl-reference.md (hunk 9 above)
Comment (sarahhaggarty) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:49]: What do you think about:
A _target_ is a part of the build responsible for compiling, testing, and packaging a piece of software for
one of the supported platforms. Kotlin provides targets for each platform so you can instruct Kotlin to compile code for that specific target. To learn more about how to  configure these targets, see [Set up targets for Kotlin Multiplatform](multiplatform-set-up-targets.md).

Comment (zamulla) [reply docs/topics/multiplatform/multiplatform-dsl-reference.md:49]: Yep, this way it's immediately clear that it's about configuration and not setting up new ones.
